CXX-877 add example for nested bson document

Machyne opened the pull request.

acmorrow commented on examples/bsoncxx/builder_stream.cpp:
```diff
@@ -58,4 +58,17 @@ int main(int, char**) {
+    // There is a special concatenate_doc helper to add all keys and corresponding values from one
+    // document into another.
+    using bsoncxx::builder::concatenate_doc;
+    doc << concatenate_doc{myQuery.view()};  // `doc` now has "foo": "bar".
```
Maybe the comment would be better if it showed the actual document. "now has" could mean a few different things.

acmorrow commented on examples/bsoncxx/builder_stream.cpp:
```diff
@@ -58,4 +58,17 @@ int main(int, char**) {
+    using bsoncxx::builder::concatenate_doc;
+    doc << concatenate_doc{myQuery.view()};  // `doc` now has "foo": "bar".
+
+    // To nested an existing bsoncxx::document::value into a builder stream, you can create a
```
nest

acmorrow commented on examples/bsoncxx/builder_stream.cpp:
```diff
@@ -58,4 +58,17 @@ int main(int, char**) {
+    // the value in.
+    bsoncxx::document::value nestedValue = document{} << "nested" << true << finalize;
+    document topLevelDoc{};
+    topLevelDoc << "subDoc1" << types::b_document{nestedValue.view()} << "subDoc2" << open_document
```
Show what the resulting document looks like here as well

acmorrow commented:
LGTM mod nits.

bazineta commented:
This example uses concatenate_doc, but there's also bsoncxx::builder::stream::concatenate; it'd be nice to have that also described, as it's not clear when or if one would be preferred over the other.

hanumantmk commented:
actually, I'm not sure we should ever be using concatenate_doc or concatenate_array directly.

The point of the free function concatenate overloads is that they'll do the right thing, and that you won't have to repeat yourself in mentioning array/doc.  I guess the one advantage is that concatenate_doc(array::view{}) will fail, where concatenate(array::view{}) would succeed (assuming the array view was elsewhere in the code and the type signature was less obvious).

I'm not sure how much you get out of that though

Machyne commented:
Thanks for the feedback! I fixed the typo, added comments to show what the actual documents look like, and switched to using the free function for `concatenate`. PTAL! :)

hanumantmk commented:
lgtm
